### ISSUE TITLE
Make raycasts support collision boxes as well

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#12673]
## Problem

Currently, raycasts can only "collide" with selection boxes but not collision boxes.

## Solutions

Add a new parameter or parameters to the raycast functions to control which type of "boxes" the raycast collides with (so: selection box only, collision box only, or both).

## Background

I noticed this limitation when I was reworking the parachute in Repixture. I use raycasts to predict whether the parachute object would collide with anything before actually spawning it. This works for most nodes but it fails for fences which have an overhigh collision box (1.5 nodes high) but a smaller selection box (1 node high).

**Comment by appgurueu** [comment]
>  I use raycasts to predict whether the parachute object would collide with anything before actually spawning it.

What you actually want is checking whether a box would collide with any other boxes which is exactly what https://github.com/appgurueu/disable_build_where_they_stand/blob/master/init.lua. does using only Lua; it was quite a hassle to implement (I had to implement all possible node collisionbox types, but that code resides on modlib, keeping dbwts simple).

**Comment by sfan5** [comment]
Like appguru said instead of extending raycasts to do this what you really want is pretty much invoking the engine's collision detection once, basically calling `collisionMoveSimple`.

**Comment by Wuzzy2** [comment]
Ignoring my specific use case, is this feature request OK? If not, close.

**Comment by Desour** [comment]
There are other use-cases for this than parachutes.
I.e. if I want to make an arrow, or some other projectile, I'd want to do a small raycast in each server-step, from old pos to new pos. This should use collision boxes.

**Comment by sfan5** [comment]
> I'd want to do a small raycast in each server-step, from old pos to new pos.

This is exactly what collision detection does, we should be exposing that.

> beam weapons obviously. pew pew

technically this is an usecase for an infinitely small beam but also collision detection can cover this.

**Comment by kno10** [comment]
For mob line-of-sight tests, raycasting *collision* boxes makes more sense than *selection boxes* IMHO.
The old `core.line_of_sight` stops on first hit, but what if we have airlike nodes that should not block line of sight?

**Comment by kno10** [comment]
> (Side note: Much of this can be implemented in pure Lua given that we do expose an API to get node collision boxes efficiently.)

Mineclonia [already has this in Lua](https://codeberg.org/mineclonia/mineclonia/src/commit/0939ca0a9689ced29726fa266fcd6016fbd78d44/mods/ENTITIES/mcl_mobs/movement.lua#L170), but (A) I don't think every mod should maintain copies of this, and (B) it supposedly is more efficient using the raycast API than iterating over all the nodes in lua and repeated retrieving the collision boxes each.
Also you are more likely to see mismatches between mods and the engine, for example due to different numerical precision used.

> I disagree. I'd say it's the exact opposite: Selection boxes per definition are more likely to coincide with what a mob would _see_. Suppose you have a mob standing in tall grass. The grass has no collision box, it can be walked through without issue. But it can not be _seen_ through, it has a selection box. You can't just punch someone through the grass, you would just punch the grass.

I disagree. First of all, some textures are simply opaque in large areas, this involves grass and flowers. Its not a solid wall, you can very well see someone behind this, you can punch someone standing inside and often even behind (if you arm is long enough), and you can very well fire an arrow through tall grass because it tends to be not that dense (in particular at the top!)

*Selection* boxes may even be oversized to simply allow selecting the object more easily. This is fine if you are selecting something (and that is another use case), but it is not appropriate for targeting or visibility.

For example mesecon wires have a much larger selection box than collision box if I am not mistaken. Now for line of sight checks (or the ability to shoot through), the collision box is the one to use IMHO, unless it is the player trying to select the mesecon wire.

> That said if a mob knew the location of its target via some sixth sense and just wants to know whether a physical entity (like an arrow) could penetrate the grass, the answer would of course be yes and a "collision box raycast" would help determine that. (I think the best example are probably gun mods. I should be able to shoot someone through grass.)

> > The old `core.line_of_sight` stops on first hit
> 
> Reasons like this are why you should use a raycast and then apply your own filtering logic to the pointed things instead.

That is what we are asking for...

> **TL;DR:** Raycasts are **not** using the wrong boxes, but there **are** legitimate use cases for them to be using collision boxes.

So this should be added as a *parameter*, right? That is what the issue proposes.

**Comment by appgurueu** [comment]
> Mineclonia [already has this in Lua](https://codeberg.org/mineclonia/mineclonia/src/commit/0939ca0a9689ced29726fa266fcd6016fbd78d44/mods/ENTITIES/mcl_mobs/movement.lua#L170)

That's nice.

> (A) I don't think every mod should maintain copies of this

Maybe you misunderstood me: I agree that an engine feature makes sense.

But in general, two important points to note:

* The engine is not responsible for deduplicating everything modders might need: Library mods should and do exist. (In fact, I see a point to be made that popular library mods have a shot at being turned into engine features for good reason as part of the process.)
* The possibility that it *is* possible in pure Lua with modest effort in my opinion lowers the priority significantly - it essentially becomes a convenience feature.

> and (B) it supposedly is more efficient using the raycast API than iterating over all the nodes in lua and repeated retrieving the collision boxes each

Most likely, which is another reason why an engine feature makes sense.

> Also you are more likely to see mismatches between mods and the engine, for example due to different numerical precision used.

I don't think numerical precision inconsistences are going to be of much importance (especially given that Lua uses doubles) and with a proper implementation I can't really think of another mismatch.

> First of all, some textures are simply opaque in large areas, this involves grass and flowers.

Issues like this are unlikely to be resolved by either selection or collisionboxes: Both are just approximations of the textured meshes that get shown to the user (but for different purposes). I'd argue that ideally, they should coincide (and this is what I would strive for in a game of mine).

> you can very well see someone behind this

It depends on the node. There definitely are nodes (maybe tall grass isn't the best example of this) which you are not - or very little - supposed to be able to see through, but which you *are* supposed to be able to walk (and shoot) through. For a more explicit example, think thick curtains to be walked through.

Selection boxes are obviously the better approximation in examples like this one.

> but it is not appropriate for targeting or visibility

The collision box is per definition for *physical collisions*. It may very well be smaller than what's visible in order to give the player and objects more space to walk. It may on the contrary be larger than what is visible to restrict movement more (e.g. fences as raised in the issue).

The selection box is for a (visual) notion of "am I pointing at this".

**Neither** is an exact "am I looking at this", a visual bounding box (for example there could also be a purely visual, nonpointable node).

> you can punch someone standing inside and often even behind (if you arm is long enough)

In the real world, sure (though actually not so sure about jungle grass), but in the Luanti world, you can't. You'd just punch the grass instead.

> For example mesecon wires have a much larger selection box than collision box if I am not mistaken

I think oversized selection boxes are something games should probably only do in extreme cases (for extremely small nodes) but that's another discussion.

(Note also that in this issue, Wuzzy presents the opposite case: His fences have a *higher* collision box than selection box. So I think either assumption of "selection box contains collision box" and "collision box contains selection box" is *wrong* in general.)

> So this should be added as a parameter, right?

That's one option, yes. (To be specific, I think at this point we might want to add a new raycast API with a single table parameter.)

(I'm not convinced of the "collision and selection boxes" mode proposed though. If modders really want that, they could just sorted merge iterate two separate raycasts in parallel.)
